robolectric/src/main/java/org/robolectric/Robolectric.java

```diff
@@ -378,4 +385,84 @@ public static Scheduler getBackgroundThreadScheduler() {
   public static void flushBackgroundThreadScheduler() {
     getBackgroundThreadScheduler().advanceToLastPostedRunnable();
   }
+
+  @SuppressWarnings({"NewApi", "unchecked"})
+  private static <T extends Service> T instantiateService(Class<T> serviceClass, Intent intent) {
+    if (RuntimeEnvironment.getApiLevel() >= P) {
+      final LoadedApk loadedApk = getLoadedApk();
+      AppComponentFactory factory = getAppComponentFactory(loadedApk);
+      if (factory != null) {
+        try {
+          return (T)
+              factory.instantiateService(
+                  loadedApk.getClassLoader(), serviceClass.getName(), intent);
+        } catch (ReflectiveOperationException e) {
+          Logger.debug("Failed to instantiate Service using AppComponentFactory", e);
+        } catch (ClassCastException ignored) {
+          // androidx.core.app.CoreComponentFactory can create CompatWrapper instance that may
+          // differ from the requested class.
+        }
+      }
+    }
+    return ReflectionHelpers.callConstructor(serviceClass);
+  }
+
+  @SuppressWarnings({"NewApi", "unchecked"})
+  private static <T extends ContentProvider> T instantiateContentProvider(Class<T> providerClass) {
+    if (RuntimeEnvironment.getApiLevel() >= P) {
+      final LoadedApk loadedApk = getLoadedApk();
+      AppComponentFactory factory = getAppComponentFactory(loadedApk);
+      if (factory != null) {
+        try {
+          return (T)
+              factory.instantiateProvider(loadedApk.getClassLoader(), providerClass.getName());
+        } catch (ReflectiveOperationException e) {
+          Logger.debug("Failed to instantiate ContentProvider using AppComponentFactory", e);
+        } catch (ClassCastException ignored) {
+          // androidx.core.app.CoreComponentFactory can create CompatWrapper instance that may
```

Comment:
nit: make the comment more generic, e.g.:

/ / check that factory.instantiateProvider actually returns an instance of T